### PR TITLE
Allow promise libraries

### PR DESCRIPTION
**Pull request by gaggle**
I wanted this library to return Q promises for my own project, with this change user can pass in a custom constructor via `.Promise` property.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -18,4 +18,6 @@ aws.Request.prototype.promise = function() {
   });
 };
 
+aws.Promise = undefined;
```
Comment (jonasfj) [inline index.js]: Why not just default it to: `require('promise');`

**Comment by jonasfj** [comment]
Curious, what is the benefit to Q promises... Isn't the whole point of the promise spec that we don't need to care about the implementation?

File: index.js (hunk 1 above)
Comment (gaggle) [inline index.js]: Well, I didn't want to load `promise` library unless necessary. Maybe that's overkill, but I figured it's one less dependency to worry about if bring my own library.

**Comment by gaggle** [comment]
Native Promises do not have all the sugar syntax that Q, Bluebird, etc. provide (e.g. `.thenResolve`). Sugar is so nice :)

**Comment by gaggle** [comment]
Rebased to latest upstream

**Comment by jonasfj** [comment]
This undoes PR #3 which moved from `that` to `.bind`, but otherwise it looks good to me...

@jhford, you also use this library quiet extensively, any thoughts?

File: index.js [hunk 2]
```diff
@@ -1,20 +1,23 @@
...
 var aws = require('aws-sdk');
-var Promise = require('promise');
 
 // XXX: This is a terrible hack but it should not break anything in horrible
 // ways.
 aws.Request.prototype.promise = function() {
-  return new Promise(function(accept, reject) {
-    this.on('complete', function(response) {
+  var that = this;
+  var Prom = aws.Promise || require('promise');
```
Comment (jhford) [inline index.js]: This defaults to using a module 'promise' if the aws.Promise property doesn't exist instead of the language's Promise class.  Could we instead do 

``` javascript
aws.Promise || Promise || require('promise');
```

File: index.js [hunk 3]
```diff
@@ -1,20 +1,23 @@
 // taken from https://github.com/taskcluster/taskcluster-queue/blob/0dd2b45b860836a2ef6c651c5499c9abfb5670ad/utils/aws-sdk-promise.js
 var aws = require('aws-sdk');
-var Promise = require('promise');
 
 // XXX: This is a terrible hack but it should not break anything in horrible
 // ways.
 aws.Request.prototype.promise = function() {
-  return new Promise(function(accept, reject) {
-    this.on('complete', function(response) {
+  var that = this;
```
Comment (jhford) [inline index.js]: is there any reason that we couldn't bind the function instead of using that?

File: index.js [hunk 4]
```diff
@@ -1,20 +1,23 @@
...
         accept(response);
       }
     });
-    this.send();
-  }.bind(this));
+    that.send();
+  });
 };
 
+aws.Promise = undefined;
```
Comment (jhford) [inline index.js]: In case `aws-sdk` ever grows a `.Promise` property, could we do a check and throw?

``` javascript
if (aws.Promise) {
    throw new Error('Refusing to overwrite aws library property');
} else {
    aws.Promise = undefined;
}
```

I guess if we did this, we could also use this as the place where we set the fallback/default for the promise library instead of doing it every single `.promise()` call.

File: index.js (hunk 3 above)
Comment (jonasfj) [inline index.js]: No, I think this is based on a revision before PR #3 which landed recently...

File: index.js (hunk 2 above)
Comment (jonasfj) [inline index.js]: Nice idea, newer nodes versions will have promise built in (node 5 does)...

File: index.js (hunk 4 above)
Comment (jonasfj) [inline index.js]: hmm, I think this is overkill... if aws introduces a promise property and `promise()` method etc, (who knows) this library will break... There is no stopping that.. We're monkey-patching here :)

Comment (jhford) [inline index.js]: yes, i'm saying that I want it to fail loud and fast when we get bitten by monkey-patching, instead of silently just altering behaviour, or in a way with really hard to parse error messages.

Comment (jonasfj) [inline index.js]: good point,

**Comment by gaggle** [comment]
Nice suggestions. Something like this?

File: index.js [hunk 5]
```diff
@@ -1,11 +1,11 @@
 // taken from https://github.com/taskcluster/taskcluster-queue/blob/0dd2b45b860836a2ef6c651c5499c9abfb5670ad/utils/aws-sdk-promise.js
 var aws = require('aws-sdk');
-var Promise = require('promise');
 
 // XXX: This is a terrible hack but it should not break anything in horrible
 // ways.
 aws.Request.prototype.promise = function() {
-  return new Promise(function(accept, reject) {
+  var Prom = aws.Promise || Promise || require('promise');
```
Comment (jonasfj) [inline index.js]: In talking to @jhford on irc... This will cause `ReferenceError` when `Promise` isn't defined...

We should probably just drop the `|| Promise ||` part and fallback to `require('promise')`.

Comment (jonasfj) [inline index.js]: Sorry... we're really bike-shedding this :)

Comment (jhford) [inline index.js]: I thought the idea was to switch to `|| global.Promise || window.Promise ||`?  It would really be nice to use the Language Promise class

Comment (jonasfj) [inline index.js]: but is `global` available in the browser or will it give us a `ReferenceError`... Hmm, maybe browserify will ensure `global` works is defined in the browser...

Comment (jhford) [inline index.js]:
``` javascript
if (global.Promise) {
  alert('have global object');
} else {
  alert('do not have it');
}
```

put through browserify running on node 4 results in this js: https://jsfiddle.net/meee4hub/ which alert's with having a global object.

Let's just check for `global.Promise` and use that.  There's no reason to depend on an external promise library when they're built into the language as of the current supported version (node 4)

**Comment by gaggle** [comment]
Now with `aws.Promise || global.Promise || window.Promise || require('promise');`.

Since we have to rely on browserify functionality do we need the `window.Promise` part?

**Comment by jhford** [comment]
We've deprecated this library now.  The upstream aws-sdk-js library has promises by default now, and you can spec which promise library to use with them.
